synapse/storage/databases/main/relations.py

```diff
@@ -671,12 +664,21 @@ async def get_bundled_aggregations(
         if not self._msc1849_enabled:
             return {}
 
-        # TODO Parallelize.
-        results = {}
+        # event ID -> bundled aggregation in non-serialized form.
+        results: Dict[str, Dict[str, Any]] = {}
+
+        event_ids = [event.event_id for event in events]
+
+        # Fetch any edits.
+        edits = await self._get_applicable_edits(event_ids)
+        for event_id, edit in edits.items():
+            results.setdefault(event_id, {})[RelationTypes.REPLACE] = edit
+
+        # Fetch other relations per event.
         for event in events:
             event_result = await self._get_bundled_aggregation_for_event(event)
             if event_result is not None:
-                results[event.event_id] = event_result
+                results.setdefault(event.event_id, {}).update(event_result)
 
```

Comment:
When I was reading this, I thought "why should edits be special here?". I guess the point is that newer edits overrule older ones. But newer e.g. emoji reactions accumulate with older ones---the behavious is different?

Reply:
Sorry, I'm not following. Are you asking why this is `update(...)` instead of setting a value? It's because `_get_bundled_aggregation_for_event` returns a dictionary of event ID -> aggregation type -> aggregation, while `_get_applicable_edits` returns event ID -> aggregation. We want to combine the edits into that dictionary and do ti by updating. We could also do the generic bundling first and then inject the edit into it (if that's clearer). That might do weird things with caching though and I would need to investigate!

Reply:
Sorry, I wasn't doing a good job of expressing myself. I can see that you need to do a setdefault-update dance _here_ because we may have already added an edit for `event.event_id` in the previous loop. That bit's clear enough to me.

I was more wondering about the big picture. Why do we handling one kind of relation (edits on +672--675) separately to other kinds of relations (+677--681)?

Reply:
Maybe the changes in #11752 will help with understanding the long term approach?

Reply:
I think I see now. I was just trying to work out if edits were somehow an "inherently different" kind of relation to the other kinds. (I.e. is the mental model different). But emphasising that this is a performance fix by reducing the number of DB queries we do makes sense!